### PR TITLE
Shuttle Explorer

### PR DESCRIPTION
**Pull request by Aurel300**
Added mechanism to emit annotated schedules from Shuttle. Added VS Code extension "Shuttle Explorer" to visualise such annotated schedules.

- [x] `cfg` flag to disable annotations completely (to avoid performance cost of checking a thread-local repeatedly)
- [ ] document and clean up extension

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/annotations/mod.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+        pub(crate) fn record_semaphore_try_acquire(object_id: ObjectId, num_permits: usize, successful: bool) {
+            record_event(AnnotationEvent::SemaphoreTryAcquire(object_id, num_permits, successful));
+        }
+
+        pub(crate) fn record_semaphore_release(object_id: ObjectId, num_permits: usize) {
+            record_event(AnnotationEvent::SemaphoreRelease(object_id, num_permits));
+        }
+
+        pub(crate) fn record_task_created(task_id: TaskId, future: bool) {
```
Comment (jorajeev) [inline src/annotations/mod.rs:303]: nit: rename argument to `is_future`

File: src/annotations/mod.rs [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+                Lazy::new(|| Regex::new(r"([0-9]+): ([^\n]+)\n +at (\./src/[^:]+):([0-9]+):([0-9]+)\b").unwrap());
+            with_state(move |state| {
+                let task_id = state.last_task_id.expect("no last task ID?"); // unwrap_or(crate::current::me());
+                                                                             /*if let Some(last_task_id) = state.last_task_id {
+                                                                                 assert_eq!(task_id, last_task_id);
+                                                                             }*/
+
+                let task_id_num = usize::from(task_id);
+                assert!(task_id_num < state.tasks.len());
+                state.tasks[task_id_num].first_step = state.tasks[task_id_num].first_step.min(state.events.len());
```
[the remark this answers is not in the document]
Comment (Aurel300) [reply src/annotations/mod.rs:173]: This `min` will only do something once, yes. However, it should be the first step of *this* task, not its creator, so we can't do it when the task is created (because its first step will either be later, or it might even never take a step).

File: src/annotations/mod.rs [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+                state.tasks[task_id_num].last_step = state.tasks[task_id_num].last_step.max(state.events.len());
+
+                use std::backtrace::Backtrace;
+                let bt = Backtrace::capture(); // TODO: config to disable
```
[the remark this answers is not in the document]
Comment (Aurel300) [reply src/annotations/mod.rs:177]: The option to disable is to disable backtraces. If we want this to be configurable separately from `RUST_BACKTRACE`, we can use `force_capture` and our own flag. I added a comment.

**Review by jorajeev — DISMISSED** [review]

File: shuttle-explorer/README.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+# Shuttle Explorer
+
+Timeline visualiser for [Shuttle](https://github.com/awslabs/shuttle).
+
+## Building
+
+To use the extension, it is currently required to build it from source. Pre-requisites include:
```
Comment (sarsko) [inline shuttle-explorer/README.md:7]: Do you know how much hassle it is to make an extension available in the VS Code store?

File: shuttle-explorer/README.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+- VS Code 1.92.0 or later
+- `node` 18.20 and `npm` 10.5 (other versions are probably fine too)
+- [esbuild Problem Matcher extension](https://marketplace.visualstudio.com/items?itemName=connor4312.esbuild-problem-matchers), installed either globally or in the Shuttle Explorer workspace. **Without this extension the builds will fail silently!**
```
Comment (sarsko) [inline shuttle-explorer/README.md:11]: Why do we need this?

File: src/annotations/mod.rs [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+        }
+
+        impl Serialize for TaskId {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::ser::Serializer,
+            {
+                usize::from(*self).serialize(serializer)
+            }
+        }
```
Comment (sarsko) [inline src/annotations/mod.rs:160]: Nit-ish:

Should all the serde stuff be moved somewhere else and put behind a feature flag "serde" (and "annotation" would enable "serde" as well)? It doesn't belong here and it also might be useful for other stuff to have a way to get serde for the types.

File: src/annotations/mod.rs [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+                // TODO: for now, "local to the project" is detected based on
+                //       the path starting with `./src/`. Find an alternative
+                //       way to do this?
+                // The following regex matches frames with local paths. We rely
+                // on the string format of the backtrace because there is no
+                // better API. At the time of writing, even the unstable feature
+                // `backtrace_frames` does not provide a better interface.
+                // https://doc.rust-lang.org/std/backtrace/struct.BacktraceFrame.html
+                static RE: Lazy<Regex> =
+                    Lazy::new(|| Regex::new(r"([0-9]+): ([^\n]+)\n +at (\./src/[^:]+):([0-9]+):([0-9]+)\b").unwrap());
```
Comment (sarsko) [inline src/annotations/mod.rs:197]: Would be nice to have a "step-by-step" comment sequence for the RegEx I think.

File: src/annotations/mod.rs [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+
+                // Whether or not the following call actually captures a backtrace
```
Comment (sarsko) [inline src/annotations/mod.rs:199]: I'm finding the code below hard to follow. Would be nice with some more comments, better names, more spacing etc.

File: src/scheduler/annotation.rs [hunk 9]
```diff
@@ -0,0 +1,47 @@
+use crate::annotations::{record_random, record_schedule, start_annotations, stop_annotations};
+use crate::runtime::task::{Task, TaskId};
+use crate::scheduler::{Schedule, Scheduler};
+
+/// An `AnnotationScheduler` wraps an inner `Scheduler` and enables the
+/// creation of an annotated schedule (for use with Shuttle Explorer).
+/// Without enabling the feature `annotation`, this wrapper does nothing.
+///
+/// Importantly, only one `AnnotatedScheduler` instance should exist at a time.
```
Comment (sarsko) [inline src/scheduler/annotation.rs:9]: Nit: `AnnotationScheduler`

File: src/sync/mutex.rs [hunk 10]
```diff
@@ -217,3 +217,17 @@ impl<T: Display + ?Sized> Display for MutexGuard<'_, T> {
         (**self).fmt(f)
     }
 }
+
+impl<T> crate::annotations::WithName for &Mutex<T> {
+    fn with_name_and_kind(self, name: Option<&str>, kind: Option<&str>) -> Self {
+        (&self.semaphore).with_name_and_kind(name, kind.or(Some("shuttle::sync::Mutex")));
```
Comment (sarsko) [inline src/sync/mutex.rs:223]: Why is this `kind.or`ed and the one below is not?

Comment (Aurel300) [reply src/sync/mutex.rs:223]: The one below is only forwarding the call to this implementation.

File: shuttle-explorer/README.md (hunk 5 above)
Comment (Aurel300) [reply shuttle-explorer/README.md:11]: As explained: without the "problem matcher", VS Code does not really understand that a background task is a background task. Some additional context [here](https://stackoverflow.com/a/49034830). This extension is also recommended by the VS Code guide for using `esbuild` for building extensions, [here](https://code.visualstudio.com/api/working-with-extensions/bundling-extension#run-esbuild).
